examples/platform/linux/AppMain.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <iostream>
+#include <thread>
+
+#include <platform/CHIPDeviceLayer.h>
+#include <platform/PlatformManager.h>
+
+#include <app/server/OnboardingCodesUtil.h>
+#include <app/server/Server.h>
+#include <core/CHIPError.h>
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+#include <setup_payload/SetupPayload.h>
+#include <support/CHIPMem.h>
+#include <support/RandUtils.h>
+
+#include "Options.h"
+
+using namespace chip;
+using namespace chip::Inet;
+using namespace chip::Transport;
+using namespace chip::DeviceLayer;
+
+namespace {
+void EventHandler(const chip::DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    (void) arg;
+    if (event->Type == chip::DeviceLayer::DeviceEventType::kCHIPoBLEConnectionEstablished)
+    {
+        ChipLogProgress(DeviceLayer, "Receive kCHIPoBLEConnectionEstablished");
+    }
+}
+} // namespace
+
+int ChipLinuxAppInit(int argc, char ** argv)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    err = chip::Platform::MemoryInit();
+    SuccessOrExit(err);
+
+    err = ParseArguments(argc, argv);
+    SuccessOrExit(err);
+
+    err = chip::DeviceLayer::PlatformMgr().InitChipStack();
+    SuccessOrExit(err);
+
+    ConfigurationMgr().LogDeviceConfig();
+    PrintOnboardingCodes(chip::RendezvousInformationFlag::kBLE);
+
+#if defined(PW_RPC_ENABLED)
+    chip::rpc::Init();
+    std::cerr << "PW_RPC initialized." << std::endl;
+#endif // defined(PW_RPC_ENABLED)
+
+    chip::DeviceLayer::PlatformMgrImpl().AddEventHandler(EventHandler, 0);
+
+    chip::DeviceLayer::ConnectivityMgr().SetBLEDeviceName(nullptr); // Use default device name (CHIP-XXXX)
+
+#if CONFIG_NETWORK_LAYER_BLE
+    chip::DeviceLayer::Internal::BLEMgrImpl().ConfigureBle(LinuxDeviceOptions::GetInstance().mBleDevice, false);
+#endif
+
+    chip::DeviceLayer::ConnectivityMgr().SetBLEAdvertisingEnabled(true);
+
+#if CHIP_DEVICE_CONFIG_ENABLE_WPA
+    if (LinuxDeviceOptions::GetInstance().mWiFi)
+    {
+        chip::DeviceLayer::ConnectivityMgrImpl().StartWiFiManagement();
+    }
+#endif // CHIP_DEVICE_CONFIG_ENABLE_WPA
+
+#if CHIP_ENABLE_OPENTHREAD
+    if (LinuxDeviceOptions::GetInstance().mThread)
+    {
+        SuccessOrExit(err = chip::DeviceLayer::ThreadStackMgrImpl().InitThreadStack());
+        std::cerr << "Thread initialized." << std::endl;
```

Comment:
Shouldn't we use the CHIP logging subsystem here?

Reply:
Yes, however, when taking look at other apps, we use the log api provided by platform directly. And when building real apps, the library other than CHIP may also print their log. Maybe we can provide a log module called "CHIPApp" for apps that want to print the logs of calling CHIP apis, or introduce some logging framework to Linux/Posix so we can avoid the use of writting logs to stderr directly?

Reply:
I guess you're right.  There are cases to be made on either side.  Yes, you can consider libCHIP one of many libraries, all of which have their own logging.  And I don't think we need a separate App logging interface.  In this context, I suppose calling to the CHIP logger simply gives some useful context and log line prefixes.

As an aside, the Linux platform is logging to stdout, not stderr, which itself seems like a poor practice.